File: JS/Map/test/map.test.js
var assert = require('assert') //Use var, const gave error
const MapCRDT = require('../map')

describe('MapCRDT', function () {
    it('should perform operations and maintain correct state', async function () {
        const map1 = new MapCRDT(1)
        const map2 = new MapCRDT(2)

        await map1.add("a", 1)
        await map1.add("b", 3)
        await map1.add("c", 5)

        await map2.add("p", 2)
        await map2.add("q", 4)
        await map2.add("r", 6)

        await map1.delete("d")
        await map2.delete("r")

        await map1.update("c", 7)
        await map2.update("r", 8)

        map1.print()
        map2.print()


        const b1 = map1.toMarshal()
        // console.log("b1: ", b1)
        const b2 = map2.toMarshal()

        const [rid1, updates1] = MapCRDT.fromMarshalData(b1)
        const [rid2, updates2] = MapCRDT.fromMarshalData(b2)

        const map3 = new MapCRDT(3)
        const map4 = new MapCRDT(4)

        await map3.merge(rid1, updates1)
        await map4.merge(rid2, updates2)

        map3.print()
        map4.print()

        assert.deepStrictEqual(map1.getValue(), map3.getValue(), "map1 and map3 are not equal")
        assert.deepStrictEqual(map2.getValue(), map4.getValue(), "map2 and map4 are not equal")
    })
})

File: Go/Map/map_test.go
package main

import (
	"reflect"
	"testing"
)

func TestMap(t *testing.T) {

	map1, map2 := NewMap(1), NewMap(2)

	map1.Add("a", 1)
	map1.Add("b", 3)
	map1.Add("c", 5)

	map2.Add("p", 2)
	map2.Add("q", 4)
	map2.Add("r", 6)

	map1.Delete("d")
	map2.Delete("r")

	map1.Update("c", 7)
	map2.Update("r", 8)

	map1.Print()
	map2.Print()

	b1 := map1.ToMarshal()
	// fmt.Println("b1: ", b1)
	b2 := map2.ToMarshal()

	rid1, updates1 := FromMarshalData(b1)
	rid2, updates2 := FromMarshalData(b2)

	map3, map4 := NewMap(3), NewMap(4)

	map3.Merge(rid1, updates1)
	map4.Merge(rid2, updates2)

	map3.Print()
	map4.Print()

	if !reflect.DeepEqual(map1.Values(), map3.Values()) {
		t.Errorf("map1 and map3 values are not the same")
	}

	if !reflect.DeepEqual(map2.Values(), map4.Values()) {
		t.Errorf("map2 and map4 values are not the same")
	}

}
